File: leetcode/company-wise/apple/0036_valid_sudoku.cpp
// LC 36. Valid Sudoku | Medium | Apple
#include <vector>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        bool rows[9][9]={}, cols[9][9]={}, boxes[9][9]={};
        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;
                int d = board[i][j]-'1', b = (i/3)*3+j/3;
                if (rows[i][d]||cols[j][d]||boxes[b][d]) return false;
                rows[i][d]=cols[j][d]=boxes[b][d]=true;
            }
        return true;
    }
};

File: leetcode/company-wise/apple/0036_valid_sudoku.py
# LC 36. Valid Sudoku | Medium | Apple
class Solution:
    def isValidSudoku(self, board):
        rows, cols, boxes = [set() for _ in range(9)], [set() for _ in range(9)], [set() for _ in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.': continue
                d = board[i][j]; b = (i//3)*3 + j//3
                if d in rows[i] or d in cols[j] or d in boxes[b]: return False
                rows[i].add(d); cols[j].add(d); boxes[b].add(d)
        return True
